### DISCUSSION TITLE
Alternatives to github after 2FA

### DISCUSSION THREAD

**Discussion by ghost** [community/community#68581] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I never had any problem with passwords ever anywhere. I already minded a lot the introduction of tokens. Not everyone maintains a security relevant repo.
Since 2FA bothers me even more, I'm looking into alternatives. 
I see gitlab, which looks not as nice to me and I heard about bitbucket.

What else is out there?

**Comment by dev-hafiz** [comment]
Alternatives to github after 2FA, I never had any problem with passwords ever anywhere. I already minded a lot the introduction of tokens. Not everyone maintains a security relevant repo.
Since 2FA bothers me even more, I'm looking into alternatives.
I see gitlab, which looks not as nice to me and I heard about bitbucket.

What else is out there? give the answer as well

ChatGPT

2 / 2

It's understandable that adjusting to new security measures can be bothersome, especially when you haven't faced issues with passwords before. If you're exploring alternatives to GitHub, here are a few options:

GitLab:

GitLab is a popular alternative that offers both a cloud-based service and a self-hosted solution. It provides features similar to GitHub and has a built-in continuous integration system.
Bitbucket:

Bitbucket, owned by Atlassian, supports both Git and Mercurial. It offers free private repositories and integrates well with other Atlassian tools like Jira.
SourceForge:

SourceForge has been around for a long time and supports Git, Mercurial, and SVN. It provides a variety of features for project management.
GitKraken:

GitKraken is not just a Git GUI client; it also offers a Git hosting service. It has a visually appealing interface and supports collaboration features.
Beanstalk:

Beanstalk is a Git and SVN hosting service that focuses on simplicity and ease of use. It provides features like deployment tools and code review.
Codeberg:

Codeberg is a community-maintained, open-source Git hosting service. It's built on Gitea, an alternative to GitLab and GitHub.

**Reply by VIPLightning** [reply]
Gitlab also uses 2FA

**Reply by pkr-sadx** [reply]
GitLab doesn't enforce 2FA and it also has email 2FA as an option.

**Reply by VIPLightning** [reply]
@pkr-sadx They do enforce it once you sign up. I remember doing that last week and it's asking for email and phone number. But of course if you're really into Gitlab, you could also give them your credit card information
![image](https://github.com/community/community/assets/18080517/61a0df42-e23a-4c0b-97e9-e0b65137f0a5)

**Comment by ghostinhershell** [comment]
Hey there!  👋

Thanks for posting in the GitHub Community, @MaestroGlanz! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category. The Accessibility category is a place for our community to discuss and provide feedback on the digital accessibility of GitHub products. Digital accessibility means that GitHub tools, and technologies, are designed and developed so that people with disabilities can use them.

With that being said, you can learn more about our decision to require 2FA for all users who contribute code to Github.com in our blog post  preceding the rollout here: [Raising the bar for software security: next steps for GitHub.com 2FA](https://github.blog/2022-12-14-raising-the-bar-for-software-security-next-steps-for-github-com-2fa/). We also have a comprehensive FAQ post that should also help clear up any questions you have on the topic. You can find that here: [Two Factor Authentication Frequently Asked Questions](https://github.com/orgs/community/discussions/68240).

I've gone ahead and moved this to the correct category for you. Good luck!

**Comment by kuroodo** [comment]
I went for GitLab. Even has an option to import your GitHub repos. I say try it out.

Bitbucket is also one I've seen people recommend

**Comment by AwokeKnowing** [comment]
I recommend everyone leave github

They are now just become bullies and another Microsoft app. 

They are unfit to host the open source community's work. Now they don't even let you decide how to authenticate. they force you to tie your account to SMS or some app. 

Move on.  

Github has lost it's way.Forcing 2fa against dev's will is EXACTLY the kind of behavior that made MS infamous. Why kill github? why don't you get it? MAKE 2FA OPTIONAL

**Comment by nmontalva** [comment]
Another example of policies enforced and implemented from places with other crime rates. The only thing accomplsihed by enforced 2FA is promotion of phone stoling.

**Comment by VIPLightning** [comment]
This is Microsoft's doing and 2FA adds potentially greater security risk. That's pretty much what's happened to 100 million epic games users that were hacked by the Russians. But I digress. I'd switch and migrate all my work to [Codeberg](https://codeberg.org/) or other git websites before December 15 at 0:00 (UTC) arrives and shit hits the fan. Because when it happens, Microsoft pretty much made this website a high valuable target to hit and Microsoft's gonna do what's only in their best interest and not for consumers or developers.

And as for @stemle-test1, this is a bot user who dislikes the community against the 2FA just to dislike. Nice try Mike Stemle

**Reply by nmontalva** [reply]
IT may be a security good practice. The best practice is not allowing any login attempt whatsoever, but that make the account useless. The point is that you need to balance between disallowing unauthorised access but also not impeding normal usage.
2FA is problematic for travelling abroad, if you don't want to use a phone, if your phone is stolen or lost, etc.
This is not only a problem of 2FA, is a problem with computer security in general: Improves safeguards against malicious access at the cost of also blocking intended access. There is obviously a trade-off that is not very well accounted for.

**Reply by VIPLightning** [reply]
Sorry Mr. Mike Stemle. But you're not getting away that fast deleting this comment.

> 2FA is a security best practice. Please cease this foolishness.

@manchicken Maybe in your own little world it's best practice. However, when there are vulnerabilities to be patched out and your SSH private keys get stolen, don't be crying. There's people who even lost or had their phones stolen and were locked out of their accounts. Just because you don't agree with something doesn't make it foolish.

**Reply by nmontalva** [reply]
My point is that the best *security* practice is also the worst *accessibility* practice, and there is no effort to balance between both.
YubiKey, FIDO2, etc, are just other options for an already faulty dogma: the doctrine known as "something you have". It is faulty because having something is a temporal status and not a condition of identity. Security systems are useless if they restrict intended usage instead of unauthorized usage.

**Reply by nmontalva** [reply]
I'd love to hear why I'm wrong, but the "CV card" is just and ad-hominem rationale. 5 km/h would be wonderful for preventing road accidents but is completely impractical.

**Reply by VIPLightning** [reply]
@manchicken

Sure, let's force everybody into MFA and just ignore everybody else's concerns. Because that's sure gonna make the community a much safer and better place. That's alright. I'm offered a 3-day grace period before I'm locked out of my account. The moment github tells me I'm no longer allowed to do anything to my account and access my data because I refused to use 2FA is the moment I stick my middle finger high at the team behind it and tell them to stick it. all of my work migrated to codeberg and Sourceforge without any BS or headaches. This community was great while it lasted. I'd understand if this was closed source and it was meant for business. Github is just an open-source platform built on top of git for general masses and the dev community, not some government employee website. But what do I know? Linking websites to convince others to use 2FA to give others the false sense of security and ignoring the concerns is the true definition of a strawman argument. 

**And as for the two users that disliked my OP post, it appears that they both are the same people except one of them is using a bot tool. I'll let you guess. Nice try Mike Stemle.**

**Reply by nmontalva** [reply]
Well, I don't know what happened to his posts, but indeed he suggested an interesting read: https://cheatsheetseries.owasp.org/cheatsheets/Multifactor_Authentication_Cheat_Sheet.html, intentionally or not, they lay down exactly the multiple issues with 2FA, and anyone who has ever used it can share experiences that add to the list of disadvantages.
Security departments are assessed in terms of breaches they have had, not considering how many actual users have been locked out, so these kind of policies are just an artificial way to improve their performance figures.

**Comment by andy-rs6** [comment]
It seems like you're exploring alternatives to traditional authentication methods like passwords, tokens, and 2FA for managing security in repositories. Here are some alternative authentication methods and version control platforms you might consider:

SSH Keys:

Instead of using passwords, you can use SSH keys for authentication. It provides a secure way to connect to your repositories without the need for passwords.
OAuth:

Some platforms, including GitHub, GitLab, and Bitbucket, support OAuth for authentication. This allows you to use third-party identity providers like Google or GitHub to authenticate users.

**Reply by pkr-sadx** [reply]
Can you stop using discussions as a test ground for your ChatGPT bots please? It's getting annoying.

**Comment by JanHitako** [comment]
I don't have any security sensitive projects at all.

This forced mesure either: 

- add a lot of unncessary complication to me logging into my github accounts if done correctly (having multiple backups, more than 2 of authentification methods) , 
- or if done simply by using just 1 or 2 types of 2FA expose me to the risk of getting me permanently locked of my account and my projects without any way of recovery.

That is because except for the sms authentification method, the other authentification methods, are reliant on you having specific devices (app on the phone, windows hello, some browser extension on the laptop, whatever). Considering where I live, phones and laptops are very common targets of stealing (and I already had my phone and my laptop stolen at the same time before from my backpack) making this such an easy way to get my account lost.

And the sms 2FA method, the only way that can mitigate my risk of getting locked permanently of my account by the way doesn't work for me. I tried multiple times in the span of a week with 2 different phone numbers, and never received, not even once, any SMS from github. 

I have other services with SMS 2FA linked to those numbers and never got any trouble receiving them. And I'm pretty sure it's a reliablity issue from github part since the amount of people experiencing the same issues is astonishing, many going back to the time where this wasn't forced on users years ago.

Point being, I really tried to deal with this new forced measure at first, but the more I look into it, the more I think it introduce more risk than benefits for my needs, therefore I'm moving away. Goodbye Github.

**Comment by HeavyMetalCookies** [comment]
I make games . And a lot of our code bases are open source . **I don't have time for this BS.**
The only time I've ever permanently lost access to an account was because of 2FA .
**GitLab** sounds good to me. I don't need anything more than backup functionality .
-KanjiCoder

**Reply by VIPLightning** [reply]
Gitlab also enforces 2FA. But Codeberg, Bitbucket, and Sourceforge are the only 3 I know that doesn't force users to use 2FA
